Mathlib/Data/Bool/Basic.lean

```diff
@@ -81,8 +81,14 @@ theorem eq_iff_eq_true_iff {a b : Bool} : a = b ↔ ((a = true) ↔ (b = true))
   cases a <;> cases b <;> simp
 
 -- Porting note: new theorem
-theorem beq_eq_decide_eq {α} [DecidableEq α]
-    (a b : α) : (a == b) = decide (a = b) := rfl
+/- Even though `DecidableEq α` implies an instance of (`Lawful`)`BEq α`, we keep the seemingly
+redundant typeclass assumptions so that the theorem is also applicable for types that have
+overridden this default instance of `LawfulBEq α` -/
+theorem beq_eq_decide_eq {α} [BEq α] [LawfulBEq α] [DecidableEq α]
```

Comment:
We might want to change this to use `{BEq α}` so that we can actually rewrite with it even if it's not the canonical `BEq` instance. That limits being able to rewrite with it in reverse though.

Reply:
Is this a common pattern, to have theorems take typeclasses as  implicit arguments to be unified rather than inferred in the usual way?

Reply:
I think this is slowly becoming more common in Lean 4 because of changes in the elaborator. IIRC, in Lean 3 instances that are found by unification are never (re-)synthesized, so there's no check that it's indeed the canonical instance.

In this case I foresee rewriting in either direction so probably it's best to keep everything an instance parameter.

Reply:
Here's the comment in the Lean prelude: https://github.com/leanprover/lean4/blob/master/src/Init/Prelude.lean#L810-L814

I think we need to update our understanding for how to handle instance arguments. It seems unavoidable needing to write specialized versions for both the forwards and backwards directions, at least until perhaps we can figure out a good RFC to get simp to automatically specialize lemmas for us...

I'm pointing this out for this particular theorem because I've run into the issue of non-canonical BEq instances, and it would be great if this lemma could apply to normalize things. I suppose I can wait until the next time this affects me and then propose a PR with the change then.